subsys/bluetooth/mesh/settings.c

```diff
@@ -1354,6 +1401,54 @@ static void store_pending_mod(struct bt_mesh_model *mod,
 	}
 }
 
+void bt_mesh_clear_va(struct label *l)
+{
+#if CONFIG_BT_MESH_LABEL_COUNT > 0
+	char path[20];
+	int err;
+
+	snprintk(path, sizeof(path), "bt/mesh/Va/%x", l->addr);
+	err = settings_delete(path);
+
+	if (err) {
+		BT_ERR("Failed to clear Virtual Address 0x%04x", l->addr);
+	} else   {
+		BT_DBG("Cleared Virtual Address 0x%04x", l->addr);
+	}
+#endif
+}
+
+#if CONFIG_BT_MESH_LABEL_COUNT > 0
+static void store_pending_va(struct label *l)
+{
+	struct va_val label;
+	char path[20];
+	int err;
+
+	if (!atomic_test_and_clear_bit(l->flag, BT_MESH_VA_CHANGED)) {
+		return;
+	}
+
+	if (lab->ref == 0) {
+		bt_mesh_clear_va(lab);
+		return;
+	}
+
+	snprintk(path, sizeof(path), "bt/mesh/Va/%x", l->addr);
```

Comment:
> that's an issue other places in the stack as well

Where? At least for incoming messages the full virtual label UUID is used as part of the decryption process, which would fail if the 16-bit version had false-positive match. I.e. the message wouldn't be passed up to the model layer in that case.

Reply:
The incoming messages are only decrypted for the first virtual label UUID that matches the 16 bit address, but if two virtual label UUIDs share the same 16 bit address, we'll never try decrypting with the second. 

Chances of this are pretty microscopic, but it's very hard to control and detect the issue for users, as the 16 bit address part of the UUID is an internal detail in the transport layer. The current problem can be fixed later without too much hassle, but if we start doing this in the persistent storage, it becomes a more permanent problem.

Reply:
@trond-snekvik agreed, thanks for elaborating